jmx/jmx.go

```diff
@@ -171,14 +169,35 @@ func openConnection(config *connectionConfig) error {
 		return err
 	}
 
+	go func() {
+		scanner := bufio.NewScanner(cmdError)
+		scanner.Buffer([]byte{}, jmxLineBuffer)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				break
+			}
+			scanner.Scan()
+
+			line := scanner.Text()
+			if strings.HasPrefix(line, "WARNING") {
+				cmdWarn <- fmt.Sprintf("cannot read nrjmx stderr: %s [proc-state: %s] (line: %s)", err, cmd.ProcessState, line)
+			}
+		}
+	}()
+
 	if err = cmd.Start(); err != nil {
 		return err
 	}
 
 	go func() {
 		if err = cmd.Wait(); err != nil {
```

Comment:
I just faced this problem in the last week: you don't have to invoke cmd.Wait until you are sure that the standard error is closed, or cmd.Wait will actually close the standard error and you will loose error logs.

The approach for the above code should be something like:

```go
canCloseOutput := sync.WaitGroup{}
canCloseOutput.Add(1)
go func() {  // you could add a similar goroutine if you want to capture stdout this way
  defer canCloseOutput.Done()
  scanner := bufio.NewScanner(cmdError)
  scanner.Buffer([]byte{}, jmxLineBuffer)
  for scanner.Scan() {
    line := scanner.Text()
    if strings.HasPrefix(line, "WARNING") {
      cmdWarn <- fmt.Sprintf("cannot read nrjmx stderr: %s [proc-state: %s] (line: %s)", err, cmd.ProcessState, line)
    }
  }
}
var waitAndClose = func() {
  if err = cmd.Wait(); err != nil {
  if err != nil {
    cmdExitErr <- fmt.Errorf("nrjmx error: %s [proc-state: %s]", err, cmd.ProcessState)
  }
}
once := sync.Once{}
defer func() {
	canCloseOutput.Wait()
	once.Do(waitAndClose)
}()
// if the task is manually cancelled, waitAndClose is invoked explicitly
go func() {
	<-ctx.Done()
	once.Do(waitAndClose)
}()
```

Reply:
Nice one, but instead of having to close the pipe before calling Wait what is required is to drain the pipe channel: https://golang.org/pkg/os/exec/#Cmd.StderrPipe

Reply:
The above code does not close the pipe. It is automatically closed by the command when it has been drained. In the link you send:
> it is incorrect to call Wait before all reads from the pipe have completed. For the same reason, it is incorrect to use Run when using StderrPipe. See the StdoutPipe example for idiomatic usage.

But if you are only waiting for the stderr to be closed, maybe the above code can be simplified to:
```go
go func() {  // you could add a similar goroutine if you want to capture stdout this way
  scanner := bufio.NewScanner(cmdError)
  scanner.Buffer([]byte{}, jmxLineBuffer)
  for scanner.Scan() {
    line := scanner.Text()
    if strings.HasPrefix(line, "WARNING") {
      cmdWarn <- fmt.Sprintf("cannot read nrjmx stderr: %s [proc-state: %s] (line: %s)", err, cmd.ProcessState, line)
    }
  }
  if err = cmd.Wait(); err != nil {
    cmdExitErr <- fmt.Errorf("nrjmx error: %s [proc-state: %s]", err, cmd.ProcessState)
  }
}

Reply:
We cannot do wait on std reads as current API is so crappy there's global vars keeping the runtime context while execution is scattered btw `Open` (trigger cmd `Wait`) and `Query` reading the std outputs.